keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,191 @@
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scailability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+### Non-Goals
+
+- Serve consistent reads from the watch cache for etcd 3.3 or earlier
+
+## Proposal
+
+Allow the etcd minor version to be specified in the kube-apiserver `--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  For backward compatibility, if `--storage-backend=etcd3` is provided, it will treat this as `etcd3.0`.
+
+Since etcd progress notify is only available in etcd 3.4. We would only serve consistent reads from cache if `--storage-backend` is set to `etcd3.4` or higher.
+For etcd 3.3-, we would continue to serve consistent reads from etcd.
+
+Guard this optmization behind `WatchCacheConsistentReads` feature gate, and only server consistent reads from cache if it is enabled.
+
+When an consistent read request is received and the watch cache is enabled:
+- Use a range(limit=0) request to get the current revision from etcd for the resource type being served. Etcd can serve this type of request efficiently, and the resulting revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Request a progress notify from etcd on the etcd watch that keeps the cache up-to-date. This is needed because the watch only receives updates for the range of the keyspace being watched. So updates for KVs outside of that range are not observed by the watch, and so the highest revision observed by the watch can be lower than the current revision even if the watch is up-to-date. A successful progress notify solve this by sending an event that will the highest revision observed to the >= the current revision. The cache is then guaranteed to to sufficiently up-to-date to serve a request at the current revision.
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+Optional: In addition to progress notify, also configure etcd to send regular
+progress events all on watches, every 10-100 milliseconds or so. If this is
+works, it could reduce the need to need to send progress notify requests to
+etcd, since watch cache progress would already kept largely up-to-date
+by the progress events etcd proactively sent.
+
+Optional: For some or all of the etcd progress watch events, also create a
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability).
+
```

Comment:
One more option @jpbetz and myself were talking about offline yesterday is the following:
- imagine that we introduce a new API to etcd that will be along those lines:
"Request: send my a response (containing id of this request) via watch once this request will be linearized and process via quorum"

This would allow us to send such request and simply serve response from watch cache on getting appropriate watch event.
It's obviously much more work on etcd side, but I would add that here as potential future improvements.

Reply:
Added to alternatives